# Assignment 4: Pipelines and Hyperparameter Tuning (32 total marks)
### Due: November 22 at 11:59pm

### Name: Nur-Alhuda Ali

### In this assignment, you will be putting together everything you have learned so far. You will need to find your own dataset, do all the appropriate preprocessing, test different supervised learning models and evaluate the results. More details for each step can be found below.

### You will also be asked to describe the process by which you came up with the code. More details can be found below. Please cite any websites or AI tools that you used to help you with this assignment.

## Import Libraries

In [1]:
import numpy as np
import pandas as pd

## Step 1: Data Input (4 marks)

Import the dataset you will be using. You can download the dataset onto your computer and read it in using pandas, or download it directly from the website. Answer the questions below about the dataset you selected. 

To find a dataset, you can use the resources listed in the notes. The dataset can be numerical, categorical, text-based or mixed. If you want help finding a particular dataset related to your interests, please email the instructor.

**You cannot use a dataset that was used for a previous assignment or in class**

In [2]:
# Import dataset (1 mark)

# If not installed, use --> pip install ucimlrepo
from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
diabetes = fetch_ucirepo(id=891) 
  
# data (as pandas dataframes) 
X = diabetes.data.features 
y = diabetes.data.targets 
  
# metadata 
print(diabetes.metadata) 
  
# variable information 
print(diabetes.variables) 


{'uci_id': 891, 'name': 'CDC Diabetes Health Indicators', 'repository_url': 'https://archive.ics.uci.edu/dataset/891/cdc+diabetes+health+indicators', 'data_url': 'https://archive.ics.uci.edu/static/public/891/data.csv', 'abstract': 'The Diabetes Health Indicators Dataset contains healthcare statistics and lifestyle survey information about people in general along with their diagnosis of diabetes. The 35 features consist of some demographics, lab test results, and answers to survey questions for each patient. The target variable for classification is whether a patient has diabetes, is pre-diabetic, or healthy. ', 'area': 'Health and Medicine', 'tasks': ['Classification'], 'characteristics': ['Tabular', 'Multivariate'], 'num_instances': 253680, 'num_features': 21, 'feature_types': ['Categorical', 'Integer'], 'demographics': ['Sex', 'Age', 'Education Level', 'Income'], 'target_col': ['Diabetes_binary'], 'index_col': ['ID'], 'has_missing_values': 'no', 'missing_values_symbol': None, 'year_

### Questions (3 marks)

1. (1 mark) What is the source of your dataset?

https://archive.ics.uci.edu/dataset/891/cdc+diabetes+health+indicators

2. (1 mark) Why did you pick this particular dataset?

I picked this dataset because diabetes runs in my family and I think it would be cool to create a model that I could test on me and my family members.

3. (1 mark) Was there anything challenging about finding a dataset that you wanted to use?

No, there wasn't. I knew I wanted to use a classification dataset for diabetes since it runs in my family.


## Step 2: Data Processing (5 marks)

The next step is to process your data. Implement the following steps as needed.

In [3]:
# Take a look at size of dataset and first few rows to look at features
print("Size of Feature Matrix: ", X.shape)
X.head()

Size of Feature Matrix:  (253680, 21)


HighBP  HighChol  CholCheck  BMI  Smoker  Stroke  HeartDiseaseorAttack  \
0       1         1          1   40       1       0                     0   
1       0         0          0   25       1       0                     0   
2       1         1          1   28       0       0                     0   
3       1         0          1   27       0       0                     0   
4       1         1          1   24       0       0                     0   

   PhysActivity  Fruits  Veggies  ...  AnyHealthcare  NoDocbcCost  GenHlth  \
0             0       0        1  ...              1            0        5   
1             1       0        0  ...              0            1        3   
2             0       1        0  ...              1            1        5   
3             1       1        1  ...              1            0        2   
4             1       1        1  ...              1            0        2   

   MentHlth  PhysHlth  DiffWalk  Sex  Age  Education  Income  
0        18        15         1    0    9          4       3  
1         0         0         0    0    7          6       1  
2        30        30         1    0    9          4       8  
3         0         0         0    0   11          3       6  
4         3         0         0    0   11          5       4  

[5 rows x 21 columns]

In [4]:
print("Size of Target Vector: ", y.shape)
y.head()

Size of Target Vector:  (253680, 1)


Diabetes_binary
0                0
1                0
2                0
3                0
4                0

In [5]:
# Get general statistics for the data.
X.describe()

HighBP       HighChol      CholCheck            BMI  \
count  253680.000000  253680.000000  253680.000000  253680.000000   
mean        0.429001       0.424121       0.962670      28.382364   
std         0.494934       0.494210       0.189571       6.608694   
min         0.000000       0.000000       0.000000      12.000000   
25%         0.000000       0.000000       1.000000      24.000000   
50%         0.000000       0.000000       1.000000      27.000000   
75%         1.000000       1.000000       1.000000      31.000000   
max         1.000000       1.000000       1.000000      98.000000   

              Smoker         Stroke  HeartDiseaseorAttack   PhysActivity  \
count  253680.000000  253680.000000         253680.000000  253680.000000   
mean        0.443169       0.040571              0.094186       0.756544   
std         0.496761       0.197294              0.292087       0.429169   
min         0.000000       0.000000              0.000000       0.000000   
25%         0.000000       0.000000              0.000000       1.000000   
50%         0.000000       0.000000              0.000000       1.000000   
75%         1.000000       0.000000              0.000000       1.000000   
max         1.000000       1.000000              1.000000       1.000000   

              Fruits        Veggies  ...  AnyHealthcare    NoDocbcCost  \
count  253680.000000  253680.000000  ...  253680.000000  253680.000000   
mean        0.634256       0.811420  ...       0.951053       0.084177   
std         0.481639       0.391175  ...       0.215759       0.277654   
min         0.000000       0.000000  ...       0.000000       0.000000   
25%         0.000000       1.000000  ...       1.000000       0.000000   
50%         1.000000       1.000000  ...       1.000000       0.000000   
75%         1.000000       1.000000  ...       1.000000       0.000000   
max         1.000000       1.000000  ...       1.000000       1.000000   

             GenHlth       MentHlth       PhysHlth       DiffWalk  \
count  253680.000000  253680.000000  253680.000000  253680.000000   
mean        2.511392       3.184772       4.242081       0.168224   
std         1.068477       7.412847       8.717951       0.374066   
min         1.000000       0.000000       0.000000       0.000000   
25%         2.000000       0.000000       0.000000       0.000000   
50%         2.000000       0.000000       0.000000       0.000000   
75%         3.000000       2.000000       3.000000       0.000000   
max         5.000000      30.000000      30.000000       1.000000   

                 Sex            Age      Education         Income  
count  253680.000000  253680.000000  253680.000000  253680.000000  
mean        0.440342       8.032119       5.050434       6.053875  
std         0.496429       3.054220       0.985774       2.071148  
min         0.000000       1.000000       1.000000       1.000000  
25%         0.000000       6.000000       4.000000       5.000000  
50%         0.000000       8.000000       5.000000       7.000000  
75%         1.000000      10.000000       6.000000       8.000000  
max         1.000000      13.000000       6.000000       8.000000  

[8 rows x 21 columns]

In [6]:
# Clean data (if needed)

# Check for null values.
print(X.dtypes)
print("\nNumber of nulls in feature matrix:\n", X.isnull().sum())
print("\nNumber of nulls in target vector: ", y.isnull().sum())

HighBP                  int64
HighChol                int64
CholCheck               int64
BMI                     int64
Smoker                  int64
Stroke                  int64
HeartDiseaseorAttack    int64
PhysActivity            int64
Fruits                  int64
Veggies                 int64
HvyAlcoholConsump       int64
AnyHealthcare           int64
NoDocbcCost             int64
GenHlth                 int64
MentHlth                int64
PhysHlth                int64
DiffWalk                int64
Sex                     int64
Age                     int64
Education               int64
Income                  int64
dtype: object

Number of nulls in feature matrix:
 HighBP                  0
HighChol                0
CholCheck               0
BMI                     0
Smoker                  0
Stroke                  0
HeartDiseaseorAttack    0
PhysActivity            0
Fruits                  0
Veggies                 0
HvyAlcoholConsump       0
AnyHealthcare           0
NoDocbc

In [7]:
# Split the data into training and testing sets.
from sklearn.model_selection import train_test_split
X_train, X_val, y_train, y_val = train_test_split(X, y, random_state=0)

In [8]:
# Implement preprocessing steps. Remember to use ColumnTransformer if more than one preprocessing method is needed

# For features that are binary, 0 or 1 values, no encoding is necessary.
# Apply ordinal encoding to these features since they are ordered & categorical but not binary: GenHlth, Age, Education, Income

# TRY WITHOUT SCALING FIRST.
# Using ColumnTransformer in case I end up adding a scaler if performance needs to be improved.
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OrdinalEncoder
ct = ColumnTransformer(
    [("ordenc", OrdinalEncoder(), ['GenHlth', 'Age', 'Education', 'Income'])
    ])


### Questions (2 marks)

1. (1 mark) Were there any missing/null values in your dataset? If yes, how did you replace them and why? If no, describe how you would've replaced them and why.

There were no missing values. If there were any missing values, I would have deleted all rows in the dataset that contain missing values since all but 3 features, 'BMI', 'MentHlth', and 'PhysHlth', are categorical and there is no reasonable way to fill in missing values from existing data.

2. (1 mark) What type of data do you have? What preprocessing methods would you have to apply based on your data types?

My dataset consists of 21 features, all but 7 are binary. Of these 7 features of type integer, 4, 'GenHlth', 'Age', 'Education', 'Income', are categorical - each integer represents are category of the feature. These categorical features of type integer must be encoded using OrdinalEncoding since their categories have an order to them. The remaining 3 integer type features, 'BMI', 'MentHlth', and 'PhysHlth', are numerical. Since 'BMI' typically does not have a wide range of values, I will first implement the models without scaling it. Since the values for 'MentHlth' and 'PhysHlth' represent days and are contained within a scale of 1 to 30 days, which is important for the interpretation of the features, I have decided not to scale these 2 features. The target vector is of binary type with 0 indicating no diabetes and 1 indicating pre-diabetes/diabetes; therefore, no encoding is needed.


## Step 3: Implement Machine Learning Model (11 marks)

In this section, you will implement three different supervised learning models (one linear and two non-linear) of your choice. You will use a pipeline to help you decide which model and hyperparameters work best. It is up to you to select what models to use and what hyperparameters to test. You can use the class examples for guidance. You must print out the best model parameters and results after the grid search.

In [9]:
# Implement pipeline and grid search here. Can add more code blocks if necessary
from sklearn.pipeline import Pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.ensemble import RandomForestClassifier

pipeline_lr = Pipeline([('preprocessing', ct), ('clsf', LogisticRegression(max_iter=1000))])
pipeline_kn = Pipeline([('preprocessing', ct), ('clsf', KNeighborsClassifier())])
pipeline_rf = Pipeline([('preprocessing', ct), ('clsf', RandomForestClassifier(random_state=0))])


In [10]:
# Parameter Grids
param_grid_lr = {'clsf__C': [0.001, 0.01, 0.1, 1, 10, 100]}
param_grid_kn = {'clsf__n_neighbors': [1, 3, 5, 7],
                 'clsf__weights': ['uniform', 'distance']}
param_grid_rf = {'clsf__n_estimators': [100, 200, 300],
                 'clsf__max_depth': [5, 10, 15]}

In [11]:
# Grid Search
grid_search_lr = GridSearchCV(pipeline_lr, param_grid_lr, cv=5, return_train_score=True)
grid_search_lr.fit(X_train, y_train.values.ravel())

grid_search_kn = GridSearchCV(pipeline_kn, param_grid_kn, cv=5, return_train_score=True)
grid_search_kn.fit(X_train, y_train.values.ravel())

grid_search_rf = GridSearchCV(pipeline_rf, param_grid_rf, cv=5, return_train_score=True)
grid_search_rf.fit(X_train, y_train.values.ravel())


GridSearchCV(cv=5,
             estimator=Pipeline(steps=[('preprocessing',
                                        ColumnTransformer(transformers=[('ordenc',
                                                                         OrdinalEncoder(),
                                                                         ['GenHlth',
                                                                          'Age',
                                                                          'Education',
                                                                          'Income'])])),
                                       ('clsf',
                                        RandomForestClassifier(random_state=0))]),
             param_grid={'clsf__max_depth': [5, 10, 15],
                         'clsf__n_estimators': [100, 200, 300]},
             return_train_score=True)

In [12]:
print("Logistic Regression Results:")
print("Best parameters: {}".format(grid_search_lr.best_params_))
print("Best cross-validation score: {:.3f}".format(grid_search_lr.best_score_))

print("\nK-Nearest Neighbors Results:")
print("Best parameters: {}".format(grid_search_kn.best_params_))
print("Best cross-validation score: {:.3f}".format(grid_search_kn.best_score_))

print("\nRandom Forest Results:")
print("Best parameters: {}".format(grid_search_rf.best_params_))
print("Best cross-validation score: {:.3f}".format(grid_search_rf.best_score_))

Logistic Regression Results:
Best parameters: {'clsf__C': 0.001}
Best cross-validation score: 0.858

K-Nearest Neighbors Results:
Best parameters: {'clsf__n_neighbors': 7, 'clsf__weights': 'uniform'}
Best cross-validation score: 0.847

Random Forest Results:
Best parameters: {'clsf__max_depth': 5, 'clsf__n_estimators': 100}
Best cross-validation score: 0.862


### Questions (5 marks)

1. (1 mark) Do you need regression or classification models for your dataset?

I need classification models for my dataset to predict the target variable as 0 (non-diabetic) or 1 (pre-diabetic or diabetic).

2. (2 marks) Which models did you select for testing and why?

I selected Logistic Regression as the linear model since it is the most commmonly used model for linear classification. For non-linear models, I wanted to use either K-Nearest Neighbours or Naive Bayes since they were both new models that we never implemented in our assignments before. However, since Naive Bayes usually performs worse than linear classifiers, I eliminated it and opted for K-Nearest Neighbours instead. The other non-linear model I chose was Random Forest because it generally has a high performance. 

3. (2 marks) Which model worked the best? Does this make sense based on the theory discussed in the course and the context of your dataset?

Although the other 2 models scored very similarly, the Random Forest Classifier performed the best, with a training accuracy of 86.2% at 100 n_estimators and a max_depth of 5. This does not agree with what was discussed in the course, as we learned that larger is always better for random forests since it reduces overfitting. This discrepancy could indicate that there could be an issue with the dataset, like a class imbalance, as indicated in the source for this dataset, causing the model to become overly fitted to the majority class and perform poorly on the minority class, especially considering that the other models' training accuracies seemed to cap at the about the same number.

## Step 4: Validate Model (6 marks)

Use the testing set to calculate the testing accuracy for the best model determined in Step 3.

In [14]:
# Calculate testing accuracy (1 mark)
from sklearn.metrics import accuracy_score

best_model = grid_search_rf.best_estimator_ 
y_pred = best_model.predict(X_val)
accuracy = accuracy_score(y_val, y_pred)
print("Testing Accuracy of Random Forest Classifier: {:.3f}".format(accuracy))

Testing Accuracy of Random Forest Classifier: 0.858



### Questions (5 marks)

1. (1 mark) Which accuracy metric did you choose? 

I used the accuracy as a metric for the model's performance, as speicfied by the instructions, which simply represents how correct the model is in its predictions.

2. (1 mark) How do these results compare to those in part 3? Did this model generalize well?

The testing accuracy is 85.8%, which is ever so slightly lower than the training accuracy. This means that the model does not generalize well and could mean the model underfits the data (is a high-bias model) and needs to be more complex. However, since it is confirmed from the source of this dataset that it is imbalanced, the more likely cause is that if one class appears more than the other in the dataset, that imbalance was present in both the training and testing set, causing the testing accuracy to be almost the same as the training accuracy.

3. (3 marks) Based on your results and the context of your dataset, did the best model perform "well enough" to be used out in the real-world? Why or why not? Do you have any suggestions for how you could improve this analysis?

No, the model did not perform well enough to be used in the real world because due to the imbalance in the dataset, the model is better trained to predict the majority class than the minority one. I would improve the model by handling the class imbalance, which could be done by oversampling (randomly duplicating samples from the minority class) or decreasing the weight of the majority class and increasing the weight of the minority class using the 'class_weight' parameter when initializing the Random Forest Classifier.

## Process Description (4 marks)
Please describe the process you used to create your code. Cite any websites or generative AI tools used. You can use the following questions as guidance:
1. Where did you source your code?
1. In what order did you complete the steps?
1. If you used generative AI, what prompts did you use? Did you need to modify the code at all? Why or why not?
1. Did you have any challenges? If yes, what were they? If not, what helped you to be successful?

1. I referenced the lecture examples for preprocessing, including scaling, encoding, column transformer, tuning, and pipelines.

2. I completed the steps in order the first time. Then I went back to make test some additional classifier models and found that they produced the similar results, which led me to realise that the issue was with the imbalanced dataset.

3. I did not use generative AI for any code. However, I did use it (ChatGPT) to understand how similar training and testing accuracies could be a result of an imbalanced dataset using the prompt "How does an imbalanced classification dataset cause a model to have the same score for training and testing?"

4. The main challenge I faced was coming to understand why my model not only seemed to cap at a certain training score and went against what we learned in class, but also why its training and testing scores were virtually the same. It took testing many additional models to finally realise that the issue could be with the data quality. This led me to look into the dataset further and my suspicion that the dataset was imbalanced was confirmed by the source I got it from. ChatGPT helped me understand how an imbalanced dataset affects model performance.

## Reflection (2 marks)
Include a sentence or two about:
- what you liked or disliked,
- found interesting, confusing, challenging, motivating
while working on this assignment.


I found it challenging to understand why my model was not performing the way I expected since I hadn't come across anything like this in the course prior to this assignment. I liked that this assignment allowed me to implement every step of the machine learning process on my own and to think critically about the preprocessing steps that are necessary for my particular dataset and what models are suitable for it.